### PR TITLE
Fix: allow escaped backreferences in `no-useless-escape` (fixes #7472)

### LINKED ISSUE #7472: no-useless-escape rule is not handling regex back references correctly
It appears that the recent changes related to the no-useless-escape rule are not handling the integer back references (e.g. `\1`) correctly.

eslint: 3.9.0
node: 4.4.3
npm: 2.15.1
parser: default

.eslintrc:

```
{
  "rules": {
    "no-useless-escape": 2,
    "no-console": 0
  }
}
```

JavaScript file:

``` js
var keyword = "'the thing \"on the inside\" is what to match'";
var exactMatchPhrases = keyword.match(/(["'])(.*?)\1/g);
console.log(exactMatchPhrases);
```

**What did you expect to happen?**
No eslint errors

**What actually happened? Please include the actual, raw output from ESLint.**
2:51  error  Unnecessary escape character: \1  no-useless-escape

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

See https://github.com/eslint/eslint/issues/7472

**What changes did you make? (Give an overview)**

This updates `no-useless-escape` to allow digit characters to be escaped in regular expressions. Escaped digits can either be backreferences or octal literals, so the rule should not warn when it encounters them.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular.

### CODE REVIEW

File: lib/rules/no-useless-escape.js [hunk 1]
```diff
@@ -36,6 +36,8 @@ const VALID_STRING_ESCAPES = [
     "\r"
 ];
 
+const DIGITS = Array.from(Array(10).keys(), String);
```
Comment (platinumazure) [inline lib/rules/no-useless-escape.js]: I have to admit, I think `const DIGITS = "0123456789".split("");` is a bit more readable.

Comment (not-an-aardvark) [inline lib/rules/no-useless-escape.js]: Yeah, maybe using Array#keys is overkill for 10 elements. I'll use that instead.

File: lib/rules/no-useless-escape.js [hunk 2]
```diff
@@ -53,7 +55,8 @@ const REGEX_GENERAL_ESCAPES = new Set([
     "W",
     "x",
     "u"
-]);
+].concat(DIGITS));
```
Comment (mysticatea) [inline lib/rules/no-useless-escape.js]: FWIW, I think this can be written as `new Set("\\0123456789bcdDfnrsStvwWxu")` since a string is iterable.

Comment (not-an-aardvark) [inline lib/rules/no-useless-escape.js]: Good idea, I updated this PR to pass in strings instead of arrays.

**Comment by mysticatea** [comment]
:+1:

**Pull request merged**